Ensure Joni warning are logged at debug

jakelandis opened the pull request:
When Joni, the regex engine that powers grok emits a warning it
does so by default to System.err. System.err logs are all bucketed 
together in the server log at WARN level.  When Joni emits a warning, 
it can be extremely verbose, logging a message for each execution 
again that pattern. For ingest node that means for every document 
that is run that Grok filter. Fortunately, Joni provides a call 
back hook to push these warnings to a custom location.

This commit implements Joni's callback hook to push the Joni warning 
to the Elasticsearch server logger (logger.org.elasticsearch.ingest.common.GrokProcessor)
at debug level. A warning will be emitted on processor creation at WARN
level.  Generally these warning indicate a possible issue with 
the regular expression. Since Grok is an abstraction on top of the regex, 
these warning may not provide much value outside of a troubleshooting 
scenario.

Additionally, the documentation is updated with instructions for how 
to set the logger to debug level.

=== 

Notes:

This error is pretty easily reproduced:

for input 
```
[foo]
```
and pattern
```
.*\[.*%{SPACE}*\].*
```
before this commit result in the following warning for every document run through the pattern. 
```
[2020-05-28T11:21:31,050][WARN ][stderr                   ] [runTask-0] regular expression has redundant nested repeat operator * /.*\[.*(?:\s*)*\].*/
```

Prior to 7.7 this message was incorrectly missed in our logs due to stderr not getting routed properly... that fix: https://github.com/elastic/elasticsearch/pull/51569 now allows this error (as shown above) to persist in the ES server log possibly causing a lot of noise to the ES server log. This fix quiets that noise.

jakelandis committed:
Ensure Joni warning are logged at debug

When Joni, the regex engine that powers grok emits a warning it does so by default to System.err. System.err logs are all bucketed together in the server log at WARN level.  When Joni emits a warning, it can be extremely verbose, logging a message for each execution again that pattern. For ingest node that means for every document that is run that Grok filter. Fortunately, Joni provides a call back hook to push these warnings to a custom location.

This commit implements Joni's callback hook to push the Joni warning to the Elasticsearch server logger (logger.org.elasticsearch.grok.Grok)
at debug level. Generally these warning indicate a possible issue with the regular expression. Since Grok is an abstraction on top of the regex, these warning may not provide much value outside of a troubleshooting scenario.

Additionally, the documentation is updated with instructions for how to set the logger to debug level.

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Ingest)

jakelandis commented:
@rjernst - I requested your review mostly for introduction of log4j from a lib.  I think that is OK ?, but also had to configure the thridparty audit.

jakelandis commented on libs/grok/src/main/java/org/elasticsearch/grok/Grok.java:
```diff
@@ -101,7 +106,8 @@ private Grok(Map<String, String> patternBank, String grokPattern, boolean namedC
         byte[] expressionBytes = expression.getBytes(StandardCharsets.UTF_8);
-        this.compiledExpression = new Regex(expressionBytes, 0, expressionBytes.length, Option.DEFAULT, UTF8Encoding.INSTANCE);
+        this.compiledExpression = new Regex(expressionBytes, 0, expressionBytes.length, Option.DEFAULT, UTF8Encoding.INSTANCE,
+            message -> logger.log(Level.DEBUG, message));
```
^^ this line is only substantive change here. (it implements the callback on the message)

martijnvg replied:
Thanks Jake! I'm happy that this is an easy change.

jasontedor reviewed (changes requested):
I don't think we should add Log4j to Grok, I think consumers of Grok should push down a logger. This makes more sense to me anyway, that the logging control should occur through the feature that is using Grok (ingest, vs. ML). This means that the Grok pattern regex needs to become non-static, which I think is fine.

jakelandis commented:
> I don't think we should add Log4j to Grok

Yeah, I was unsure if we should do that.  Are the lib projects subject to semver w.r.t the java API ? (e.g.  

>  I think consumers of Grok should push down a logger

Do think SLF4J (and we bind it to Log4j upstream) or pass in something custom like what Joni does ? (or possibly just expose the Joni callback)

jasontedor commented:
>Yeah, I was unsure if we should do that. Are the lib projects subject to semver w.r.t the java API ?

No, [they are for internal use](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#L6).

>Do think SLF4J (and we bind it to Log4j upstream) or pass in something custom like what Joni does ? (or possibly just expose the Joni callback)

Neither, how about just expecting a `Consumer<String>` when constructing `Grok`?

rjernst commented:
I agree with Jason's assessment; we should not depend on log4j in a lib.

Separately, are these warnings only emitted when running Grok? Can we not get them when we compile the pattern in the first place? It seems like these should be returned to a user, not hidden at debug level. They can indicate an error, like in your example maybe a `+` should have been there after space instead of a `*` (I realize that doesn't make sense in this case, but image space is something else meaningful).

jakelandis commented:
> Separately, are these warnings only emitted when running Grok? 

Yes, as far as i can tell, the warnings are emitted at the runtime execution, not on the pattern compile.  

> Can we not get them when we compile the pattern in the first place?

Not that I am aware. 

> It seems like these should be returned to a user, not hidden at debug level

Agreed, however without some kind throttled logger (i.e. only log this at most once every 10m) I am not sure how that could be done. This is why i included the doc to help out with manually increasing the log level.  Groks is an abstraction over regex, and the resultant regex (even for basic patterns) can get so large they become almost human un-readable. I think at best with real world patterns it is a signal that something might be off, but not likely something concrete actionable from the warning.

jakelandis committed:
Revert "Ensure Joni warning are logged at debug"

This reverts commit 2f64eb74e14b74119798a661c300f63edaa350e9.

jakelandis committed:
use callback instead of direct log4j dependency

jakelandis committed:
add on construction warning

jakelandis commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/GrokProcessor.java:
```diff
@@ -48,9 +51,12 @@
         this.ignoreMissing = ignoreMissing;
+        // Joni warnings are only emitted on an attempt to match, and the warning emitted for every call to match which is too verbose
+        // so here we emit a warning (if there is one) to the logfile at warn level on construction / processor creation.
+        new Grok(patternBank, combinePatterns(matchPatterns, traceMatch), matcherWatchdog, logger::warn).match("___nomatch___");
```
this is kinda weird, but hopefully the comment explains why. This allows Joni warnings to be emitted at WARN level on PUT of the pipeline, but DEBUG for normal running. 

If we put a WARN for normal running we can flood the log (and what this PR seeks to address)

jakelandis committed:
fix import

jakelandis commented:
@rjernst @jasontedor - ready for another review. I updated to use a callback for the log message and removed the log4j dependency in the lib (I had to intro it the module). I also added a call to match on the constructor to get it to emit the warning on construction.. its kinda hacky, but the only way i can see to get the warning in the log file without flooding the log file.

rjernst commented on modules/ingest-common/build.gradle:
```diff
@@ -27,6 +27,7 @@ dependencies {
   compileOnly project(':modules:lang-painless')
   compile project(':libs:elasticsearch-grok')
   compile project(':libs:elasticsearch-dissect')
+  compile "org.apache.logging.log4j:log4j-api:${versions.log4j}"
```
This is already included as a compileOnly dep through server, so it shouldn't be needed here?

jakelandis replied:
thanks, I missed that this was already a dependency

rjernst commented on modules/ingest-common/license/log4j-api-LICENSE.txt:
```diff
@@ -0,0 +1,202 @@
+
+                                 Apache License
```
the license file should not be needed since the log4j jar is not actually bundled in this module; it is part of server

jakelandis replied:
fixed. thanks

rjernst reviewed (commented):
Looks fine. One more main question

jakelandis committed:
remove compile dependency on log4jApi and remove license

jakelandis commented:
@rjernst - mind taking another look?  I fixed the license and dependency on [https://github.com/elastic/elasticsearch/pull/57302/commits/2b2f269f49fe8e8c0d0f83392517f462543cb65b ](https://github.com/elastic/elasticsearch/pull/57302/commits/2b2f269f49fe8e8c0d0f83392517f462543cb65b)

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into route_joni_warnings_to_debug_log

rjernst reviewed (approved):
LGTM

jakelandis commented:
@rjernst thanks for the review
@jasontedor - merging - your requested change (removing the log4j dependency) has been addressed.
